### ISSUE TITLE
Salesforce Streaming API

### ISSUE THREAD

**Issue by chulkilee** [chulkilee/ex_force#9]
[Introducing Streaming API](https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/intro_stream.htm)

Salesforce Streaming API is based on Bayeux / ComeD. Unfortunately, I couldn't find a library to handle that :(

**Comment by MrDoops** [comment]
It's 7 years since the last commit, but there's an Erlang library that might work: https://github.com/antoniogarrote/erlfaye

**Comment by chulkilee** [comment]
I'm not super clear how Bayeux / Cometd / Faye relate to each other..

[Faye](https://faye.jcoglan.com/architecture.html) says:

> Faye is based on the Bayeux protocol, and is compatible with the reference implementation of Bayeux provided by the cometD project.

so I guess any Faye library can be used - I found another one in elixir:  https://github.com/messagerocket/pixie

**Comment by MrDoops** [comment]
Faye is used in [Restforce](https://github.com/restforce/restforce#streaming), the Ruby library I'm working with currently. It's known to only work with a specific version. I wonder why the implementation is so finicky? 

It [looks](https://github.com/messagerocket/pixie/issues/2) like Pixie is still on Elixir 1.3 and could probably use some updates. 

I'd love to be using Elixir for this Salesforce integration project I'm working on. Elixir makes more sense than Rails to being a sort of reactive API-consumer middle-layer which is what the use case revolves around.

**Comment by chulkilee** [comment]
@MrDoops I'm on the same boat :)

I'm not using Salesforce streaming api, but even so I'm not sure I want to implement it if there is no good library I can use. My use cases do not need to create topic dynamically, so I may create topics and subscribe with other tool (e.g. java, I assume the best support for this), and use different pubsub mechanism (e.g. kafka or whatever).

Another option is to write APEX class and code to call my api back, but that has different challenges of deployment, monitoring, logging, etc.

**Comment by MrDoops** [comment]
Salesforce itself provides the server mechanisms for the CometD/Bayeux Protocol implementation. The Pixie repo implements both the server and client portions, but it's still worth referencing. For ExForce, we'd only need to implement a client that can pub/sub and handle re-connections. Good description of how Faye implements it [here](https://faye.jcoglan.com/architecture.html).

[This](https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/using_streaming_api_client_connection.htm) documentation from Salesforce is probably the most succinct description of the implementation requirements for a Salesforce specific CometD client.

From what I can tell we'll want a persistent processes listening to channels like `meta/handshake` channel to handle messages from the server informing of re-connections, network failures, and handshake success/failure. Then maybe a dynamic supervisor to spawn listener processes to an individual topics/subscription that implement the long polling. The [Java CometD code](https://github.com/cometd/cometd/blob/master/cometd-java/cometd-java-client/src/main/java/org/cometd/client/BayeuxClient.java) is worth checking out.

**Comment by fullstackpiyush** [comment]
Hi,
How can I be helpful to this issue..
Is it resolved?

**Comment by chulkilee** [comment]
@fullstackpiyush no, it's not implemented. I'm not using this feature, so I don't have a plan to work on it - please feel free to take a look and create PR. Thanks!
